### PR TITLE
PR #10 edits

### PR DESCRIPTION
**Pull request by zenodeapp**
@bengtlofgren, here:

## Solved (imo; but test it)
**Edge cases:** 
- If the polled block is higher than the finalized block it shouldn't continue else it will mark that polled block as scraped.
- Since finalized doesn't have an initial scrape it will skip checking the very first block (START_BLOCK). Made sure to let the variable `lastScrapedBlock` have `START_BLOCK - 1` as the default fallback value.

**Bugs:**
- One import bug
- The `isRunningFinalized` variable for the second scheduler wasn't used, instead it pointed to `isRunning` of the default scheduler. 
- Block number and finalized block number were not compared correctly due to comparing a hex with a number. 

**Features**
- Added batching for receipts. Set it to 30 max to play it safe, but it's a variable so can be altered. But check whether this is fine or if we should skip this. I left the simpler approach in comments.
- Abstracted the logic for the scheduler and removed duplicate code. I use objects to keep track of whether a scheduler `isRunning` (renamed to `busy`) or not since objects can be mutated when passed as a parameter in JavaScript.

## What's left?
- Totals API endpoint.
- What exactly should the end user know about finalized/non-finalized? I feel like it could easily complicate things (even scare people away). So perhaps we should not.
- We could potentially create an initial scrape method for finalized. Would likely only be useful for testing as it would not matter for mainnet - since the scraper would already be turned on and at the latest finalized block height when we launch.

### CODE REVIEW

File: api/checkDonation.js [hunk 1]
```diff
@@ -11,7 +11,7 @@ const pool = new Pool({
   });
 
 async function findCutoffData(finalized = false) {
-  const viewTable = finalized ? 'donation_stats_finalized' : 'donations';
+  const viewTable = finalized ? 'donation_stats_finalized' : 'donation_stats';
```
Comment (bengtlofgren) [inline api/checkDonation.js:14]: Good catch!

File: api/checkDonation.js [hunk 2]
```diff
@@ -64,13 +64,13 @@ async function checkEthAddress(ethAddress, cutoffData, finalized = false) {
           ELSE 0
         END as address_eligible_eth
       FROM ${table} 
-      WHERE from_address = $1 AND (block_number < $2 OR (block_number = $2 AND tx_index < $3))
+      WHERE lower(from_address) = $1 AND (block_number < $2 OR (block_number = $2 AND tx_index < $3))
```
Comment (bengtlofgren) [inline api/checkDonation.js:67]: Nice

File: api/checkDonation.js [hunk 3]
```diff
@@ -64,13 +64,13 @@ async function checkEthAddress(ethAddress, cutoffData, finalized = false) {
...
     ),
     cutoff_tx AS (
       -- Get the cutoff transaction if it exists
       SELECT amount_eth
       FROM ${table}
-      WHERE from_address = $1 AND block_number = $2 AND tx_index = $3
+      WHERE lower(from_address) = $1 AND block_number = $2 AND tx_index = $3
```
Comment (bengtlofgren) [inline api/checkDonation.js:73]: Nice

File: api/checkDonation.js [hunk 4]
```diff
@@ -86,7 +86,7 @@ async function checkEthAddress(ethAddress, cutoffData, finalized = false) {
           0
         ) as eligible_eth
       FROM ${table} 
-      WHERE from_address = $1
+      WHERE lower(from_address) = $1
```
Comment (bengtlofgren) [inline api/checkDonation.js:89]: V nice

File: api/checkDonation.js [hunk 5]
```diff
@@ -153,15 +153,15 @@ async function checkNamadaAddress(namadaAddress, cutoffData, finalized = false)
...
+      WHERE lower(namada_key) = $1
     )
     SELECT total_eth, eligible_eth FROM address_total
   `;
 
   const MaxBigInt = BigInt('9223372036854775807');  // for block_number (BIGINT)
   const MaxInt = 2147483647;        
 
-  const result = await pool.query(query, [namadaAddress, cutoffData.cutoff_block || MaxBigInt.toString(), cutoffData.cutoff_tx_index || MaxInt]);
+  const result = await pool.query(query, [namadaAddress.toLowerCase(), cutoffData.cutoff_block || MaxBigInt.toString(), cutoffData.cutoff_tx_index || MaxInt]);
```
Comment (bengtlofgren) [inline api/checkDonation.js:164]: Super duper

File: lib/infura.mjs [hunk 6]
```diff
@@ -51,24 +52,34 @@ export async function getTransactionReceipt(txHash, retryCount = 0) {
...
+      if (finalizedBlock > blockNumber) {
+        // If the finalized block is greater than the latest polled block, it means that we missed this block, so call the function again with the latest block number.
+        return await getBlockTransactions(blockNumber, false);
+      }
+
+      if(blockNumber > finalizedBlock) {
+        // If the polled block is greater than the finalized block, we should not continue.
+        if(VERBOSE_LOGGING) log(`Block ${blockNumber}: not finalized yet.`);
+        return null;
+      }
```
Comment (bengtlofgren) [inline lib/infura.mjs:81]: This is fair, I don't think it's a bug to continue since nothing will be added if the data is already there, but this method does save on api calls

File: lib/scheduler.mjs [hunk 7]
```diff
@@ -4,123 +4,118 @@ import {
...
-        );
-      }
-    } catch (error) {
-      logError('Error in cron job:', error);
-    } finally {
-      isRunning = false;
-    }
+export function startSchedulerFinalized() {
+  cron.schedule(`*/${schedulerFinalized.interval} * * * * *`, async () => {
+    await scrapeBlockchain(schedulerFinalized);
```
Comment (bengtlofgren) [inline lib/scheduler.mjs:37]: Nice, you went ahead and did the concise way, i commend you

**Review by bengtlofgren — APPROVED** [review]

File: lib/infura.mjs (hunk 6 above)
Comment (zenodeapp) [reply lib/infura.mjs:81]: It will mark `blockNumber` in the scraped finalized blocks table as scraped which is problematic, right?

Comment (bengtlofgren) [reply lib/infura.mjs:81]: Oh yes this may be true, yup you're right
